src/Microsoft.ML.Parquet/ParquetLoader.cs

```diff
@@ -173,36 +157,122 @@ private ParquetLoader(Arguments args, IHost host, Stream stream)
                 }
 
                 _columnChunkReadSize = args.ColumnChunkReadSize;
-                InitColumns(ch, out _columnsLoaded);
+                _columnsLoaded = InitColumns(schemaDataSet);
                 Schema = CreateSchema(_host, _columnsLoaded);
             }
         }
 
+        private ParquetLoader(IHost host, ModelLoadContext ctx, IMultiStreamSource files)
+        {
+            Contracts.AssertValue(host);
+            _host = host;
+            _host.AssertValue(ctx);
+            _host.AssertValue(files);
+
+            // *** Binary format ***
+            // int: cached chunk size
+            // bool: TreatBigIntegersAsDates flag
+            // Schema of the loader (0x00010002)
+
+            _columnChunkReadSize = ctx.Reader.ReadInt32();
+            bool treatBigIntegersAsDates = ctx.Reader.ReadBoolean();
+
+            if (ctx.Header.ModelVerWritten >= 0x00010002)
+            {
+                // Load the schema
+                byte[] buffer = null;
+                if (!ctx.TryLoadBinaryStream(SchemaCtxName, r => buffer = r.ReadByteArray()))
+                    throw _host.ExceptDecode();
+                BinaryLoader loader = null;
+                var strm = new MemoryStream(buffer, writable: false);
+                loader = new BinaryLoader(_host, new BinaryLoader.Arguments(), strm);
+                Schema = loader.Schema;
+            }
+
+            // Only load Parquest related data if a file is present. Otherwise, just the Schema is valid.
+            if (files.Count > 0)
+            {
+                _parquetOptions = new ParquetOptions()
+                {
+                    TreatByteArrayAsString = true,
+                    TreatBigIntegersAsDates = treatBigIntegersAsDates
+                };
+
+                _parquetStream = OpenStream(files);
+                DataSet schemaDataSet;
+
+                try
+                {
+                    // We only care about the schema so ignore the rows.
+                    ReaderOptions readerOptions = new ReaderOptions()
+                    {
+                        Count = 0,
+                        Offset = 0
+                    };
+                    schemaDataSet = ParquetReader.Read(_parquetStream, _parquetOptions, readerOptions);
+                    _rowCount = schemaDataSet.TotalRowCount;
+                }
+                catch (Exception ex)
+                {
+                    throw new InvalidDataException("Cannot read Parquet file", ex);
+                }
+
+                _columnsLoaded = InitColumns(schemaDataSet);
+                var streamSchema = CreateSchema(_host, _columnsLoaded);
+
+                if (Schema == null)
+                {
+                    Schema = streamSchema;
+                }
+                else if (!Schema.EqualColumns(streamSchema))
+                {
+                    throw _host.Except("File schema does not match the model schema");
```

Comment:
I'm not sure I like the idea of a hard fail. Can we make this like `BinaryLoader`? From memory:

1. If there are no files, it loads the internal schema (what you are doing here more or less),
2. If there is a file, it uses that instead, and ignores the internal schema.

I might prefer to be consistent. Is there any harm with that scheme?

Reply:
If there is a hard fail though, I *might* prefer that it be more specific. That is just saying "they're not the same" isn't as actionable feedback as we can give. Maybe if it somehow printed out where it differed that would work?

---
In reply to: [201386238](https://github.com/dotnet/machinelearning/pull/472#discussion_r201386238) [](ancestors = 201386238)

Reply:
I have to objection to your request. Falling in line with BinaryLoader.

---
In reply to: [201386608](https://github.com/dotnet/machinelearning/pull/472#discussion_r201386608) [](ancestors = 201386608,201386238)

Reply:
> I have to objection

A Freudian slip!! :D :D

---
In reply to: [201761047](https://github.com/dotnet/machinelearning/pull/472#discussion_r201761047) [](ancestors = 201761047,201386608,201386238)